src/Orleans.Serialization/Buffers/ArcBufferWriter.cs

```diff
@@ -0,0 +1,1404 @@
+    private ArcBufferPage _tail;
+
+    // The offset into the first page which has been consumed already. When this reaches the end of the page, the page can be unpinned.
+    private int _readIndex;
+
+    // The total length of the buffer.
+    private int _totalLength;
+
+    /// <summary>
+    /// Gets the minimum page size.
+    /// </summary>
+    public const int MinimumPageSize = ArcBufferPagePool.MinimumPageSize;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ArcBufferWriter"/> struct.
+    /// </summary>
+    public ArcBufferWriter()
+    {
+        _readPage = _writePage = _tail = ArcBufferPagePool.Shared.Rent();
+        Debug.Assert(_readPage.ReferenceCount == 0);
+        _readPage.Pin(_readPage.Version);
+    }
+
+    /// <summary>
+    /// Gets the number of unconsumed bytes.
+    /// </summary>
+    public int Length => _totalLength - _readIndex;
+
+    /// <summary>
+    /// Adds additional buffers to the destination list until the list has reached its capacity.
+    /// </summary>
+    /// <param name="buffers">The destination to add buffers to.</param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void ReplenishBuffers(List<ArraySegment<byte>> buffers)
+    {
+        // Skip half-full pages in an attempt to minimize the number of buffers added to the destination
+        // at the expense of under-utilized memory. This could be tweaked up to increase page utilization.
+        const int MinimumUsablePageSize = MinimumPageSize / 2;
+
+        while (buffers.Count < buffers.Capacity)
+        {
+            // Only use the current page if it is greater than the minimum "usable" page size.
+            if (_tail.WriteCapacity > MinimumUsablePageSize)
+            {
+                buffers.Add(_tail.WritableArraySegment);
+            }
+
+            // Allocate a new page.
+            AllocatePage(0);
+        }
+    }
+
+    /// <inheritdoc/>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    void IBufferWriter<byte>.Advance(int count) => AdvanceWriter(count);
+
+    /// <summary>
+    /// Advances the writer by the specified number of bytes.
+    /// </summary>
+    /// <param name="count">The numbers of bytes to advance the writer by.</param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void AdvanceWriter(int count)
+    {
+        _totalLength += count;
+        while (true)
+        {
+            var amount = Math.Min(_writePage.WriteCapacity, count);
+            _writePage.Advance(amount);
+            count -= amount;
+
+            if (count == 0)
+            {
+                break;
+            }
+
+            var next = _writePage.Next;
+            Debug.Assert(next is not null);
+            _writePage = next;
+        }
+    }
+
+    /// <summary>
+    /// Resets this instance, returning all memory.
+    /// </summary>
+    public void Reset()
```

Comment:
Is this safe if there's an ArcBuffer that has a reference to this memory? We unpin the page and return it to the page pool but the ArcBuffer referencing it doesn't AFAICT know that the page may be being reused. The version check is only used during pin/unpin but not on access.